chore(npm): move tsd update to post install and add tsd version to devDe...

IgorMinar opened the pull request:
...pendencies

IgorMinar committed:
chore(npm): move tsd update to post install and add tsd version to devDependencies

IgorMinar committed:
chore(npm): add shrinkwrap to our project

IgorMinar committed:
chore(npm): remove duplicate fs-extra dependency

IgorMinar committed:
chore(travis): add npm/install-dependencies.sh step in preparation for caching

IgorMinar committed:
chore(travis): turn on travis cache for npm dependencies

IgorMinar committed:
chore(travis): pin npm used to v2.7.4

Since we are currently stuck on node 0.10 (see #1396), this will at least help us avoid
some of the npm bugs present in older version that comes with 0.10.

IgorMinar commented:
I'm seeing some proctor tests fail after upgrade to node 0.12. @juliemr could you please take a look at https://travis-ci.org/angular/angular/builds/58700821 Do you have any idea why node upgrade would cause this?

juliemr commented:
Nothing jumps out right away. The failures are all in benchpress tests, so I'm guessing it's something the benchpress framework is doing. Base Protractor is testing against node 0.12, so there are no known issues.

IgorMinar commented:
thanks @juliemr. downgrading to 0.10 does seem to solve the problem: https://travis-ci.org/angular/angular/jobs/58702783

the failure with 0.12 was pretty consistent so I'm not suspecting a flake in this case.

juliemr commented:
Here's what I've been able to figure out - with node 0.12.2, benchpress occasionally gets into a state where webdriver commands are executed out of order. Normally, it wants to load a url, find an element, and then click:

```
22:53:23.503 INFO - Executing: [get: http://localhost:8001/benchmarks/src/change_detection/change_detection_benchmark.html?numberOfChecks=900000&iterations=20])
22:53:23.503 INFO - Executing: [find elements: By.cssSelector: #baselineChangeDetectionWrites])
22:53:23.629 INFO - Done: [get: http://localhost:8001/benchmarks/src/change_detection/change_detection_benchmark.html?numberOfChecks=900000&iterations=20]
22:53:23.706 INFO - Done: [find elements: By.cssSelector: #baselineChangeDetectionWrites]
22:53:23.726 INFO - Executing: [click: 5 [[ChromeDriver: chrome on MAC (elemid)] -> css selector: #baselineChangeDetectionWrites]])
22:53:24.162 INFO - Done: [click: 5 [[ChromeDriver: chrome on MAC (elemid)] -> css selector: #baselineChangeDetectionWrites]]
```

but it gets out of order and fails with staleElementException:

```
22:53:31.699 INFO - Executing: [find elements: By.cssSelector: #instantiateDirectives])
22:53:31.699 INFO - Executing: [get: http://localhost:8001/benchmarks/src/element_injector/element_injector_benchmark.html?iterations=20000])
22:53:31.716 INFO - Done: [find elements: By.cssSelector: #instantiateDirectives] <<---- THIS HAPPENED TOO SOON
22:53:31.726 INFO - Executing: [click: 7 [[ChromeDriver: chrome on MAC (67c52459a9a095ee413c7927d8896a68)] -> css selector: #instantiateDirectives]])
22:53:31.838 INFO - Done: [get: http://localhost:8001/benchmarks/src/element_injector/element_injector_benchmark.html?iterations=20000]
22:53:32.299 WARN - Exception thrown
org.openqa.selenium.StaleElementReferenceException
```

The webdriver control flow *should* prevent this, whether Protractor is involved or not. In fact, the commands should always be `executing` followed by `done`. I'm not sure why two `executing` commands are getting queued up at once.

IgorMinar commented:
ok. thanks @juliemr. I created #1396 for this issue and will for now stick to node 0.10.

@alexeagle can you review these changes or should I find someone else?

alexeagle commented on .travis.yml:
```diff
@@ -28,12 +33,24 @@ before_install:
+  - npm install -g npm@2.7.4
+  # Check the size of caches
+  - du -sh ./node_modules || true
+  # Instal npm dependecies and ensure that npm cache is not stale
```
install

alexeagle commented on .travis.yml:
```diff
@@ -2,6 +2,11 @@ language: node_js
+
+cache:
+  directories:
+    - node_modules
```
curious: did you look into caching the typings directories yet?

IgorMinar replied:
I want to do it step by step. there are many other dirs that need to be cached

IgorMinar replied:
or rather that can be cached :)

alexeagle commented on tools/npm/clean-shrinkwrap.js:
```diff
@@ -0,0 +1,38 @@
+console.log('Cleaning shrinkwrap object');
+cleanModule(shrinkwrap, shrinkwrap.name);
+
+var cleanShrinkwrapPath = path.join(__dirname, '..', '..', 'npm-shrinkwrap.clean.json');
```
path.normalize?

IgorMinar replied:
join automatically normalizes the path

alexeagle commented on tools/npm/clean-shrinkwrap.js:
```diff
@@ -0,0 +1,38 @@
+
+var cleanShrinkwrapPath = path.join(__dirname, '..', '..', 'npm-shrinkwrap.clean.json');
+console.log('Writing cleaned to', cleanShrinkwrapPath);
+fs.writeFileSync(cleanShrinkwrapPath, JSON.stringify(sorted(shrinkwrap), null, 2) + "\n");
```
Did you see the trick where you put your function in the second arg to JSON.stringify?
https://github.com/npm/npm/issues/3581#issuecomment-28301361
Not sure if that's equivalent to what you're doing here...

IgorMinar replied:
we tried that in the past but there are several problems:

- we need the order of dependencies to be stable, as it is now, it's not so even small updates can generate huge diffs which are impossible to review. This is why the .clean.json file exists
- we need a separate file because some of those messy unstable properties which are meaningless to humans are needed for npm. don't ask me why :) but we've had an issue in angular 1 where by us cleaning up the original shrinkwrap file we made npm install the wrong thing. oops.

alexeagle commented on tools/npm/clean-shrinkwrap.js:
```diff
@@ -0,0 +1,38 @@
+function cleanModule(module, name) {
+
+  // keep `resolve` properties for git dependencies, delete otherwise
+  delete module.from;
```
syntax highlighting is confusing here, it thinks this refers to the ES6 module global. consider renaming the parameter?

alexeagle commented on tools/npm/install-dependencies.sh:
```diff
@@ -0,0 +1,16 @@
+set -e
+
+SHRINKWRAP_FILE=npm-shrinkwrap.json
+SHRINKWRAP_CACHED_FILE=node_modules/npm-shrinkwrap.cached.json
```
nice trick, this file will be copied over by travis caching so we can effectively checksum the directory...

IgorMinar replied:
yes

alexeagle commented:
this will be super helpful.

IgorMinar commented:
ok, I'll address the small comments and get this in.

IgorMinar commented:
comments address, running one more test build and then merging this in

IgorMinar committed:
chore(build): trigger npm postinstall when npm install is skipped

we temporarily need this because some of our dependencies are being installed via npm postinstall
